config/allowlist.go

```diff
@@ -54,43 +57,96 @@ type Allowlist struct {
 	// This targets the _secret_, not the content of the regex match like the
 	// Regexes slice.
 	StopWords []string
+
+	// commitMap is a normalized version of Commits, used for efficiency purposes.
+	// TODO: possible optimizations so that both short and long hashes work.
+	commitMap    map[string]struct{}
+	regexPat     *regexp.Regexp
+	pathPat      *regexp.Regexp
+	stopwordTrie *ahocorasick.Trie
 }
 
+var (
+	flagOnce         sync.Once
+	useOptimizations bool
+)
+
 // CommitAllowed returns true if the commit is allowed to be ignored.
 func (a *Allowlist) CommitAllowed(c string) (bool, string) {
-	if c == "" {
+	if a == nil || c == "" || len(a.Commits) == 0 {
 		return false, ""
 	}
-
-	for _, commit := range a.Commits {
-		if commit == c {
-			return true, c
+	if useOptimizations {
+		if _, ok := a.commitMap[strings.ToLower(c)]; ok {
+			return true, ""
+		}
+	} else {
+		for _, commit := range a.Commits {
+			if commit == c {
+				return true, c
+			}
 		}
 	}
 	return false, ""
 }
 
 // PathAllowed returns true if the path is allowed to be ignored.
 func (a *Allowlist) PathAllowed(path string) bool {
-	return anyRegexMatch(path, a.Paths)
+	if a == nil || path == "" {
+		return false
+	}
+
+	if useOptimizations {
+		if a.pathPat == nil {
+			return false
+		}
+		return a.pathPat.MatchString(path)
+	} else {
+		return anyRegexMatch(path, a.Paths)
+	}
 }
 
 // RegexAllowed returns true if the regex is allowed to be ignored.
 func (a *Allowlist) RegexAllowed(secret string) bool {
-	return anyRegexMatch(secret, a.Regexes)
+	if a == nil || secret == "" {
+		return false
+	}
+
+	if useOptimizations {
+		if a.regexPat == nil {
+			return false
+		}
+		return a.regexPat.MatchString(secret)
+	} else {
+		return anyRegexMatch(secret, a.Regexes)
+	}
 }
 
 func (a *Allowlist) ContainsStopWord(s string) (bool, string) {
+	if a == nil || s == "" || len(a.StopWords) == 0 {
+		return false, ""
+	}
+
 	s = strings.ToLower(s)
-	for _, stopWord := range a.StopWords {
-		if strings.Contains(s, strings.ToLower(stopWord)) {
-			return true, stopWord
+	if useOptimizations {
+		if m := a.stopwordTrie.MatchFirstString(s); m != nil {
+			return true, m.MatchString()
+		}
+	} else {
+		for _, stopWord := range a.StopWords {
+			if strings.Contains(s, stopWord) {
+				return true, stopWord
+			}
 		}
 	}
 	return false, ""
 }
 
 func (a *Allowlist) Validate() error {
+	flagOnce.Do(func() {
```

Comment:
Relying on an 'optional' call to `Validate` to set internal state is bad. A better idea is probably to make a `NewAllowlist(...opts)` function, but I didn't want to add noise before this change got reviewed.